test/csuite/timestamp_abort/main.c

```diff
@@ -313,26 +296,33 @@ thread_run(void *arg)
 	    td->info, td->start);
 	stable_ts = 0;
 	for (i = td->start;; ++i) {
-		if (use_ts)
-			stable_ts = __wt_atomic_addv64(&global_ts, 1);
 		testutil_check(__wt_snprintf(
 		    kname, sizeof(kname), "%" PRIu64, i));
 
 		testutil_check(session->begin_transaction(session, NULL));
 		if (use_prep)
 			testutil_check(oplog_session->begin_transaction(
 			    oplog_session, NULL));
-		/*
-		 * If not using prepared transactions set the timestamp now
-		 * before performing the operation. If we are using prepared
-		 * transactions, it must be set after the prepare.
-		 */
-		if (use_ts && !use_prep) {
-			testutil_check(__wt_snprintf(tscfg, sizeof(tscfg),
-			    "commit_timestamp=%" PRIx64, stable_ts));
-			testutil_check(
-			    session->timestamp_transaction(session, tscfg));
+
+		if (use_ts) {
+			testutil_check(pthread_rwlock_wrlock(&ts_lock));
+			stable_ts = __wt_atomic_addv64(&global_ts, 1);
+			/*
+			 * If not using prepared transactions set the timestamp
+			 * now before performing the operation. If we are using
+			 * prepared transactions, it must be set after the
+			 * prepare.
+			 */
+			if (!use_prep) {
```

Comment:
I'm concerned with how long the lock is held when doing prepared transactions. I can see that the test program writes the timestamp into the value, so wants to allocate one before doing work.

Because of two sessions/transactions are used when prepared is in use, we know that there will be a transaction using the timestamp at this point (in `session`), so the global value can't be updated. I think it's OK to release the lock here, as long as the prepare call on `oplog_session` happens prior to the `session->commit` call.

Reply:
Releasing the lock here causes a test failure : 
`[1538524024:257641][16257:0x7f945b7f6700], WT_SESSION.prepare_transaction: __wt_txn_parse_prepare_timestamp, 863: prepare timestamp 7dc0 is older than the oldest timestamp 7f87 : Invalid argument
test_timestamp_abort: FAILED: thread_run/361: session->prepare_transaction( session, tscfg): Invalid argument`
I had tried this during initial implemetation, hence moved the lock release after `oplog_session`'s commit for prepared txns.

Reply:
Even if you re-order the prepare and commit further down?

Reply:
The reason is that if this thread allocates the timestamp earlier (with the lock even) and gets N, and some other thread commits N+1 and then the timestamp-coordinator thread calls `query_timestamp` it doesn't know anything about N and N+1 is the last-committed. So oldest/stable get updated to N+1 and when this thread starts using N it gets that error.